datastream-server/src/main/java/com/linkedin/datastream/server/assignment/StickyPartitionAssignmentStrategy.java

```diff
@@ -141,6 +142,126 @@ public StickyPartitionAssignmentStrategy(Optional<Integer> maxTasks, Optional<In
     return newAssignment;
   }
 
+  /**
+   * Move a partition for a datastream group according to the targetAssignment. As we are only allowed to mutate the
+   * task once. It follow the steps
+   * Step 1) get the partitions that to be moved, and find out their source task
+   * Step 2) If the instance is the one we want to move, we find out the task which we should assign the partition
+   *         from that instance
+   * Step 3) Scan the current assignment, compute new task if the old task belongs to these source tasks or if it
+   *         is the target task we want to move to
+   *
+   * @param currentAssignment the old assignment
+   * @param targetAssignment the target assignment retrieved from Zookeeper
+   * @param partitionsMetadata the subscribed partitions metadata received from connector
+   * @return new assignment
+   */
+  public Map<String, Set<DatastreamTask>> movePartitions(Map<String, Set<DatastreamTask>> currentAssignment,
```

Comment:
instead of adding a movePartitions implementation in strategy, I was thinking if we can create an interface findNewTask/similar, which returns new task(s) to assign the new partitions to.  All the tasks modifications are taken care of outside of strategy (if the modifications are generic enough for all connector types/strategies). That way new strategies just need to implement this new findNewTask method. Default implementation of findNewTask will be just returning task with least #partitions.

Reply:
I think you mean split step 1) into findNewTask, I don't think we can benefit too much as I don't think there are other usecases. The task modifications are should also be considered as the part of assignment strategy.